File: apps/ios/M0ProtocolMockClient/Sources/M0ProtocolMockClient/Models/M0Event.swift
import Foundation

/// All M0 event variants supported by the mock client.
public enum M0Event: Codable, Equatable, Sendable {
    /// Stream connection lifecycle event.
    case connected(M0ConnectedEvent)

    /// Terminal or process output chunk.
    case outputChunk(M0OutputChunkEvent)

    /// Generic status transition event.
    case status(M0StatusEvent)

    /// Liveness pulse event.
    case heartbeat(M0HeartbeatEvent)

    private enum CodingKeys: String, CodingKey {
        case type
        case payload
    }

    private enum EventType: String, Codable {
        case connected
        case outputChunk = "output_chunk"
        case status
        case heartbeat
    }

    /// Decodes a typed M0 event from a discriminator-based payload format.
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(EventType.self, forKey: .type)

        switch type {
        case .connected:
            self = .connected(try container.decode(M0ConnectedEvent.self, forKey: .payload))
        case .outputChunk:
            self = .outputChunk(try container.decode(M0OutputChunkEvent.self, forKey: .payload))
        case .status:
            self = .status(try container.decode(M0StatusEvent.self, forKey: .payload))
        case .heartbeat:
            self = .heartbeat(try container.decode(M0HeartbeatEvent.self, forKey: .payload))
        }
    }

    /// Encodes a typed M0 event into a discriminator-based payload format.
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)

        switch self {
        case let .connected(event):
            try container.encode(EventType.connected, forKey: .type)
            try container.encode(event, forKey: .payload)
        case let .outputChunk(event):
            try container.encode(EventType.outputChunk, forKey: .type)
            try container.encode(event, forKey: .payload)
        case let .status(event):
            try container.encode(EventType.status, forKey: .type)
            try container.encode(event, forKey: .payload)
        case let .heartbeat(event):
            try container.encode(EventType.heartbeat, forKey: .type)
            try container.encode(event, forKey: .payload)
        }
    }
}

/// Payload for a connection lifecycle event.
public struct M0ConnectedEvent: Codable, Equatable, Sendable {
    /// Server session identifier.
    public let sessionID: String

    /// Last sequence accepted by the server for replay resume.
    public let acceptedLastSeq: UInt64?

    /// Creates a connected event payload.
    /// - Parameters:
    ///   - sessionID: Server session identifier.
    ///   - acceptedLastSeq: Last sequence accepted by server.
    public init(sessionID: String, acceptedLastSeq: UInt64?) {
        self.sessionID = sessionID
        self.acceptedLastSeq = acceptedLastSeq
    }

    private enum CodingKeys: String, CodingKey {
        case sessionID = "session_id"
        case acceptedLastSeq = "accepted_last_seq"
    }
}

/// Payload containing stream output data.
public struct M0OutputChunkEvent: Codable, Equatable, Sendable {
    /// Text chunk emitted by the stream.
    public let chunk: String

    /// Marks whether this chunk ends the current output frame.
    public let isFinal: Bool

    /// Creates an output chunk payload.
    /// - Parameters:
    ///   - chunk: Stream output chunk.
    ///   - isFinal: Final chunk indicator.
    public init(chunk: String, isFinal: Bool) {
        self.chunk = chunk
        self.isFinal = isFinal
    }

    private enum CodingKeys: String, CodingKey {
        case chunk
        case isFinal = "is_final"
    }
}

/// Payload describing a status transition.
public struct M0StatusEvent: Codable, Equatable, Sendable {
    /// Machine-readable status code.
    public let code: String

    /// Human-readable status description.
    public let message: String

    /// Creates a status payload.
    /// - Parameters:
    ///   - code: Machine-readable status code.
    ///   - message: Human-readable status description.
    public init(code: String, message: String) {
        self.code = code
        self.message = message
    }
}

/// Payload used to indicate stream liveness.
public struct M0HeartbeatEvent: Codable, Equatable, Sendable {
    /// Server epoch timestamp in milliseconds.
    public let serverTimeMs: UInt64

    /// Creates a heartbeat payload.
    /// - Parameter serverTimeMs: Server epoch milliseconds.
    public init(serverTimeMs: UInt64) {
        self.serverTimeMs = serverTimeMs
    }

    private enum CodingKeys: String, CodingKey {
        case serverTimeMs = "server_time_ms"
    }
}

File: apps/android/M0ProtocolMockClient/src/main/kotlin/com/commandrelay/m0protocol/models/M0Event.kt
package com.commandrelay.m0protocol.models

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.descriptors.element
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * M0 event variants encoded as a discriminator object with `type` and `payload`.
 */
@Serializable(with = M0EventSerializer::class)
public sealed class M0Event {
    /**
     * Stream connection lifecycle event.
     *
     * @property payload Connected event payload.
     */
    public data class Connected(public val payload: M0ConnectedEvent) : M0Event()

    /**
     * Terminal or process output chunk event.
     *
     * @property payload Output chunk payload.
     */
    public data class OutputChunk(public val payload: M0OutputChunkEvent) : M0Event()

    /**
     * Generic status transition event.
     *
     * @property payload Status payload.
     */
    public data class Status(public val payload: M0StatusEvent) : M0Event()

    /**
     * Liveness pulse event.
     *
     * @property payload Heartbeat payload.
     */
    public data class Heartbeat(public val payload: M0HeartbeatEvent) : M0Event()
}

/**
 * Payload for a connection lifecycle event.
 *
 * @property sessionId Server session identifier.
 * @property acceptedLastSeq Last sequence accepted by the server for replay resume.
 */
@Serializable
public data class M0ConnectedEvent(
    @SerialName("session_id")
    public val sessionId: String,
    @SerialName("accepted_last_seq")
    public val acceptedLastSeq: Long? = null,
)

/**
 * Payload containing stream output data.
 *
 * @property chunk Text chunk emitted by the stream.
 * @property isFinal Whether this chunk ends the current output frame.
 */
@Serializable
public data class M0OutputChunkEvent(
    public val chunk: String,
    @SerialName("is_final")
    public val isFinal: Boolean,
)

/**
 * Payload describing a status transition.
 *
 * @property code Machine-readable status code.
 * @property message Human-readable status description.
 */
@Serializable
public data class M0StatusEvent(
    public val code: String,
    public val message: String,
)

/**
 * Payload used to indicate stream liveness.
 *
 * @property serverTimeMs Server epoch timestamp in milliseconds.
 */
@Serializable
public data class M0HeartbeatEvent(
    @SerialName("server_time_ms")
    public val serverTimeMs: Long,
)

internal object M0EventSerializer : KSerializer<M0Event> {
    override val descriptor: SerialDescriptor =
        buildClassSerialDescriptor("M0Event") {
            element<String>("type")
            element<JsonObject>("payload")
        }

    override fun serialize(encoder: Encoder, value: M0Event) {
        val jsonEncoder = encoder as? JsonEncoder
            ?: throw SerializationException("M0Event can only be serialized by JSON")
        val json = jsonEncoder.json

        val (type, payload) = when (value) {
            is M0Event.Connected -> "connected" to json.encodeToJsonElement(M0ConnectedEvent.serializer(), value.payload)
            is M0Event.OutputChunk -> "output_chunk" to json.encodeToJsonElement(M0OutputChunkEvent.serializer(), value.payload)
            is M0Event.Status -> "status" to json.encodeToJsonElement(M0StatusEvent.serializer(), value.payload)
            is M0Event.Heartbeat -> "heartbeat" to json.encodeToJsonElement(M0HeartbeatEvent.serializer(), value.payload)
        }

        // Keep wire shape aligned with iOS and protocol docs: `{ "type": "...", "payload": { ... } }`.
        jsonEncoder.encodeJsonElement(
            buildJsonObject {
                put("type", JsonPrimitive(type))
                put("payload", payload)
            },
        )
    }

    override fun deserialize(decoder: Decoder): M0Event {
        val jsonDecoder = decoder as? JsonDecoder
            ?: throw SerializationException("M0Event can only be deserialized by JSON")
        val root = jsonDecoder.decodeJsonElement().jsonObject
        val type = root["type"]?.jsonPrimitive?.content
            ?: throw SerializationException("Missing event type")
        val payload = root["payload"]
            ?: throw SerializationException("Missing event payload")
        val json = jsonDecoder.json

        return when (type) {
            "connected" -> M0Event.Connected(json.decodeFromJsonElement(M0ConnectedEvent.serializer(), payload))
            "output_chunk" -> M0Event.OutputChunk(json.decodeFromJsonElement(M0OutputChunkEvent.serializer(), payload))
            "status" -> M0Event.Status(json.decodeFromJsonElement(M0StatusEvent.serializer(), payload))
            "heartbeat" -> M0Event.Heartbeat(json.decodeFromJsonElement(M0HeartbeatEvent.serializer(), payload))
            else -> throw SerializationException("Unsupported event type: $type")
        }
    }
}
